### PR TITLE
CA1501: Add configurability for exclude_structs

### LINKED ISSUE #3178: Prefer private/public fields in EditorConfig
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/idea/842310/scope-for-net-naming-conventions-for-editorconfig.html)._

---
<p>Is it possible to distinguish the scope for .editorconfig naming rules?</p> <p>I'd like to allow public fields in structs:</p><pre><code>public struct MyStruct {     public int StructField; } </code></pre> <p>But disallow public fields in classes:</p><pre><code>public class MyClass {     public int ClassField; } </code></pre> <p>Is this possible using .editorconfig naming rules?</p> <p>Back link: <a href="https://github.com/MicrosoftDocs/visualstudio-docs/issues/4400" target=_blank>https://github.com/MicrosoftDocs/visualstudio-docs/issues/4400</a></p>

---
#### Visual Studio Feedback System on 12/4/2019, 00:30 AM: 

Thank you for taking the time to provide your suggestion.  We will do some preliminary checks to make sure we can proceed further.  We'll provide an update once the issue has been triaged by the product team.
#### Kendra Havens [MSFT] on 1/3/2020, 06:49 PM: 

<p>Currently, you are not able to prefer private or public fields per type via the EditorConfig. In your second example for disallowing public fields in classes, would you want it to show a warning or just a message?</p>
<p>Thank you for helping Visual Studio improve!</p>

#### Luke on 1/3/2020, 11:48 PM: 

<p>I'd prefer a warning (which can be configured to be considered as an error which can prevent the project from being built).</p>

### PR DESCRIPTION
**Pull request by Evangelink**
Fix https://github.com/dotnet/roslyn-analyzers/issues/3178

Tagging @mavasani

### CODE REVIEW

File: docs/fundamentals/code-analysis/quality-rules/ca1051.md [hunk 1]
```diff
@@ -49,13 +49,29 @@ Consumers may need field access in the following situations:
...
 You can configure which parts of your codebase to run this rule on, based on their accessibility. For example, to specify that the rule should run only against the non-public API surface, add the following key-value pair to an *.editorconfig* file in your project:
 
 ```ini
 dotnet_code_quality.ca1051.api_surface = private, internal
 ```
 
-You can configure this option for just this rule, for all rules, or for all rules in this category (Design). For more information, see [Code quality rule configuration options](../code-quality-rule-options.md).
+### Exclude structs
+
+You can configure whether non-private `struct` (`Structure` in Visual Basic) fields should be analyzed.
```
Comment (mavasani) [inline docs/fundamentals/code-analysis/quality-rules/ca1051.md:70]: Nit:
```suggestion
You can exclude `struct` (`Structure` in Visual Basic) fields from being analyzed.
```

@gewarren probably has a better wording suggestion.

File: docs/fundamentals/code-analysis/quality-rules/ca1051.md [hunk 2]
```diff
@@ -49,13 +49,29 @@ Consumers may need field access in the following situations:
...
+#### Include specific API surfaces
+
 You can configure which parts of your codebase to run this rule on, based on their accessibility. For example, to specify that the rule should run only against the non-public API surface, add the following key-value pair to an *.editorconfig* file in your project:
 
 ```ini
 dotnet_code_quality.ca1051.api_surface = private, internal
 ```
 
-You can configure this option for just this rule, for all rules, or for all rules in this category (Design). For more information, see [Code quality rule configuration options](../code-quality-rule-options.md).
+### Exclude structs
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1051.md:68]:
```suggestion
#### Exclude structs
```

File: docs/fundamentals/code-analysis/quality-rules/ca1051.md [hunk 3]
```diff
@@ -49,13 +49,29 @@ Consumers may need field access in the following situations:
 
 ## Configurability
 
+This rule has the following configurable options that can be configured for just this rule, for all rules, or for all rules in this category (Design). For more information, see [Code quality rule configuration options](../code-quality-rule-options.md).
+
+### Include or exclude selected APIs
+
+Use the following options to configure which parts of your codebase to run this rule on.
+
+- [Include specific API surfaces](#include-specific-api-surfaces)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1051.md:58]:
```suggestion
- [Include specific API surfaces](#include-specific-api-surfaces)
- [Exclude structs](#exclude-structs)
```

File: docs/fundamentals/code-analysis/quality-rules/ca1051.md [hunk 4]
```diff
@@ -49,13 +49,29 @@ Consumers may need field access in the following situations:
 
 ## Configurability
 
+This rule has the following configurable options that can be configured for just this rule, for all rules, or for all rules in this category (Design). For more information, see [Code quality rule configuration options](../code-quality-rule-options.md).
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1051.md:52]:
```suggestion
This rule has the following options that can be configured for just this rule, for all rules, or for all rules in this category (Design). For more information, see [Code quality rule configuration options](../code-quality-rule-options.md).
```

**Comment by Evangelink** [comment]
@gewarren @mavasani We have the same question that is repeating across all my PRs, we have added a `### Include or exclude selected APIs` under `## Configurability` but as far as I see most of the configurability we add are actually around the inclusion and exclusion of what is anayzed so it feels a little weird to me.

I would recommend to either:
- get rid of the H3 and move up all H4
- Rename H2 `Configurability` into H3 `Include or exclude selected APIs`

Let me know what you want/like.

File: docs/fundamentals/code-analysis/quality-rules/ca1051.md [hunk 5]
```diff
@@ -49,13 +49,29 @@ Consumers may need field access in the following situations:
 
 ## Configurability
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1051.md:50]:
```suggestion
## Include or exclude APIs
```

File: docs/fundamentals/code-analysis/quality-rules/ca1051.md [hunk 6]
```diff
@@ -49,13 +49,29 @@ Consumers may need field access in the following situations:
...
 
+This rule has the following options that can be configured for just this rule, for all rules, or for all rules in this category (Design). For more information, see [Code quality rule configuration options](../code-quality-rule-options.md).
+
+### Include or exclude selected APIs
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1051.md:54]:
```suggestion
```

File: docs/fundamentals/code-analysis/quality-rules/ca1051.md [hunk 7]
```diff
@@ -49,13 +49,29 @@ Consumers may need field access in the following situations:
...
+
+Use the following options to configure which parts of your codebase to run this rule on.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1051.md:56]:
```suggestion
```

File: docs/fundamentals/code-analysis/quality-rules/ca1051.md [hunk 8]
```diff
@@ -49,13 +49,29 @@ Consumers may need field access in the following situations:
 
 ## Configurability
 
+This rule has the following options that can be configured for just this rule, for all rules, or for all rules in this category (Design). For more information, see [Code quality rule configuration options](../code-quality-rule-options.md).
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1051.md:52]:
```suggestion
This rule has options that let you configure which parts of your codebase to run the rule on. The options can be configured for just this rule, for all rules, or for all rules in this category (Design). For more information, see [Code quality rule configuration options](../code-quality-rule-options.md).
```

File: docs/fundamentals/code-analysis/quality-rules/ca1051.md [hunk 9]
```diff
@@ -49,13 +49,29 @@ Consumers may need field access in the following situations:
...
+
+### Include or exclude selected APIs
+
+Use the following options to configure which parts of your codebase to run this rule on.
+
+- [Include specific API surfaces](#include-specific-api-surfaces)
+
+#### Include specific API surfaces
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1051.md:60]:
```suggestion
### Include specific API surfaces
```

**Review by gewarren — COMMENTED** [review]
I think we should rename the Configurability H2 to "Include or exclude APIs", delete the "Include or exclude selected APIs" H3, and promote the H4s to H3s.

File: docs/fundamentals/code-analysis/quality-rules/ca1051.md [hunk 10]
```diff
@@ -47,15 +47,30 @@ Consumers may need field access in the following situations:
 - In ASP.NET Web Forms content controls.
 - When the target platform makes use of `ref` to modify fields, such as model-view-viewmodel (MVVM) frameworks for WPF and UWP.
 
-## Configurability
+## Include or exclude APIs
+
+This rule has options that let you configure which parts of your codebase to run the rule on. The options can be configured for just this rule, for all rules, or for all rules in this category (Design). For more information, see [Code quality rule configuration options](../code-quality-rule-options.md).
+
+
+
```
Comment (Evangelink) [inline docs/fundamentals/code-analysis/quality-rules/ca1051.md:55]:
```suggestion
```

**Comment by Evangelink** [comment]
Thank you @gewarren! I suggest to merge this as-is and I will do the "include" comments in a separate PR.

**Review by gewarren — APPROVED** [review]

**Pull request merged**